src/plonk_util.rs

```diff
@@ -280,7 +280,8 @@ pub(crate) fn polynomial_degree<F: Field>(
     coeffs.iter().rev().skip_while(|c| c.is_zero()).count() - 1
 }
 
-fn halo_s<F: Field>(us: &[F]) -> Vec<F> {
+// TODO: Maybe a streaming version using an `Iterator` would be faster and wouldn't require as much memory for large circuits. 
```

Comment:
I think we could also do this with 2*degree field muls by exploiting the common terms. Not sure if it's worth the work to optimize though.

Reply:
Yes I agree there should be a way to greatly optimize this. I'll add a TODO.